bestofjs/bestofjs-webui
# Sanitize input before rendering to users

## Goal
- [x] Sanitize input before rendering to users using `dangerouslySetInnerHTML` (`src/components/project-details/readme-card.tsx`)
- [x] Run prettier on the whole project

## Why
We add one more layer of security so if our database is compromised. Our users are not at stake. This is [Defense in Depth](https://www.forcepoint.com/cyber-edu/defense-depth) principle (if one layer fails, another steps up immediately to thwart an attack)

## How to test

- Update `src/mocks/data/project-readme-react.html.js` like this

```diff
const projectReadmeReact = `<div id="readme" class="md" data-path="README.md">
+<img src/onerror=alert('stole-your-token')>
  <article class="markdown-body entry-content container-lg" itemprop="text">
...
```

- Run
```
npm run start:mock
```

- Navigate to any project, for example: http://localhost:3000/projects/freecodecamp. You should see the javascript is executed if we do not `sanitize` input.

## Screenshots

<img width="816" alt="pop up in an alert window" src="https://user-images.githubusercontent.com/8603085/173858918-5e5dc09f-9dc3-4d39-ba86-b5b0c5529c3b.png">
